### ISSUE TITLE
CalDAV URL

### ISSUE THREAD

**Issue by 150d** [WLAN-Kabel/ioBroker.calendar#15]
Before claiming that something is wrong, I'd like to make sure I don't have a configuration error:

I'm trying to access a NextCloud account containing a number of calendars. What is the URL that should be entered in the adapter config?

https://servername
-> objects for one calendar are created, but no "day X", just account/color/name

https://servername/remote.php/dav/calendars/username/
-> same

https://servername/remote.php/dav/calendars/username/calendar1name
-> existing calendar (matching the URL) is renamed to the true NextCloud-name, additional calendars are discovered and created in the config. But for the new calenders, the URL is still .../calendar1name, not .../calendar2name etc. If I change the entries manually, nothing more happens. No appointments are retrieved in any calendars. In the log there are a number of errors:

(9552) TypeError: Cannot read property 'length' of undefined
(9245) TypeError: Cannot read property 'val' of undefined

For some calendars, objects named "day X" are created, for others they are not. For the "original" calendar (whose URL was originally specified and whose URL is now set for all discovered calendars), they are not.

Do I need to set a different URL?

Regards

**Comment by WLAN-Kabel** [comment]
Hey, the right one should be https://example.com/remote.php/dav/principals/users/

or

https://example.com/remote.php/dav/principals/

**Comment by 150d** [comment]
Ok, I tried https://example.com/remote.php/dav/principals/.

No change: The calendars are discovered, but no data is retrieved. The objects "day x" are not created. There is one error per (activated) calendar in the log:

TypeError: Cannot read property 'length' of undefined

**Comment by WLAN-Kabel** [comment]
Ok, I will check it later.

**Comment by WLAN-Kabel** [comment]
Ok, https://example.com/remote.php/dav/principals is correct.

Did you try to delete the calendar and add it again?

If the account has already been read in "successfully", it will not be read in again.

**Comment by 150d** [comment]
Yes, a number of times. The calendars are successfully discovered (and named correctly), but no content is transferred. I've seen the "day x" objects created before, but they were all zeros. Currently, they are not even created. I thought this might be connected to the error message - it looks like processing of the downloaded data is aborted before any content is decoded.

**Comment by WLAN-Kabel** [comment]
Please try it with the new version 1.1.1

**Comment by 150d** [comment]
No change, I'm afraid.

I specified a single calendar in the settings. After connecting the first time, all the calendars in my NextCloud account were detected. Root objects were created for all of them, but only for 2 out of the 8 I had enabled there are "day X" objects. These calendars do not have anything scheduled for the next 7 days (but other calendars also have nothing scheduled for 7 days and have not have the "day X" objects created.)

The log errors are still there:
[401] null (once only)
TypeError: Cannot read property 'length' of undefined (3 times)
TypeError: Cannot read property 'val' of undefined (1 time)

**Comment by WLAN-Kabel** [comment]
The [401] means that the request isn't authorized.

Are you using the correct credentials?

Please activate debug mode and check which calendar is trying to load.

Did you enter the password again after the update? This is necessary, otherwise the password will be read incorrectly.

**Comment by 150d** [comment]
I had removed the adapter (installed from repository) completly and installed the new one (from GitHub URL). On creating the instance, I had to redo the config from scratch (as expected.)

The credentials I believe to be correct, because otherwise the calendars wouldn't have been discovered, would they?

Which Debug mode do you mean? I'm unsure where to enable that.

**Comment by WLAN-Kabel** [comment]
You can activate debug mode for each instance in the instance overview.

 Manual (in German, you only need the pictures): https://forum.iobroker.net/assets/uploads/files/139_iobroker_debug_anleitung_f_r_adapter.pdf

**Comment by 150d** [comment]
I have DEBUG enabled now, thanks. I found the following correlations:

All calendars that show no error in the log have "day X" objects created. The calendars with error do not.
Only the first calendar shows error "401 null", all other errors are "TypeError".
Even if "day X" objects are created, there are no events in them (count = 0).
Even if a calendar doesn't have events for 7 days, it may still produce an error.
The errors are not in sequence: Even if one calendar goes through without error, the next may produce an error again, and the next again may not.

**Comment by WLAN-Kabel** [comment]
Hey, I've found a bug that triggers the 401. The error generates inconsistent login data when discovering the calendars. I will fix this asap. I will continue to look at the other errors.

**Comment by 150d** [comment]
Great! This also explains why my NextCloud instance was "delay[ing] login because of brute force attack from this IP". Originally I thought that I had activated/deactivated the adapter too often (causing ever new connection attempts).

**Comment by 150d** [comment]
I can confirm that the 401 is no longer occuring.

Unfortunatly, the other errors ("TypeError: Cannot read property 'length' of undefined") still _do_ occur, and there still are no "day X" objects created (for the two calendars I tested on.)

**Comment by WLAN-Kabel** [comment]
Nice to hear that one bug is fixed.

I will continue to look for the other bug.

Thanks for the feedback.

**Comment by WLAN-Kabel** [comment]
Could you please try the development version (installation via [https://github.com/WLAN-Kabel/ioBroker.calendar/tarball/development](https://github.com/WLAN-Kabel/ioBroker.calendar/tarball/development) in the adapter menu)

If the error isn't resolved, then please report the debug output.

**Comment by 150d** [comment]
This version is looking very much different than the previous ones:

- I did the usual install/config. On first connect, all calendars were discovered and the first one was auto-activated (as before). This first calendar did download one event (there only is a single event from today to +7). "Day X" objects were created and the single event was registered. There were no errors in the log.

- Then I activated _two_ more calendars. The log showed _one_ error ("ERROR TypeError: Cannot read property 'val' of undefined"). _One_ of the calendars had "Day X" objects created and one event downloaded, the other did not. I located the spot in debug log. Can I send this to you non-publicly (PM, email or something?)

- The calendar that failed (no objects, log error) seems to have downloaded a lot of events, apparently everything there is and ever was: E.g. events that were due in 2014 (yes, six years ago) are downloaded, at least according to the debug log. Is this expected behaviour?
